sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,208 @@
+  void setFuncPtr<detail::PiApiKind::api>(RT::PiPlugin * MPlugin,              \
+                                          decltype(&::api) FuncPtr) {          \
+    MPlugin->PiFunctionTable.api = FuncPtr;                                    \
+  }
+#include <CL/sycl/detail/pi.def>
+
+/// The PiMock class wraps an instance of a SYCL platform class,
+/// and manages all mock redefinitions for the underlying plugin.
+///
+/// Mock platform instances must not share the plugin resources with
+/// any other SYCL platform within the given context. Otherwise, mock
+/// redefinitions would also affect other platforms' behavior.
+/// Therefore, any plugin-related information is fully copied whenever
+/// a user-passed SYCL object instance is being mocked.
+/// The underlying SYCL platform must be a non-host plaftorm to facilitate
+/// plugin usage.
+///
+/// Simple usage examples would look like this:
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// unittest::PiMock Mock;
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// platform &MockP = Mock.getPlatform();
+/// /*...*/
+/// ```
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// queue Q;
+/// unittest::PiMock Mock(Q);
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// Q.submit(/* expecting mock behavior */);
+/// /*...*/
+/// ```
+class PiMock {
+public:
+  /// Default-constructs PiMock.
+  ///
+  /// Delegates to the device_selector-based constructor by passing
+  /// a default_constructor and assuming a non-host device will be
+  /// available.
+  PiMock() : PiMock(cl::sycl::default_selector{}) {}
+
+  /// Constructs PiMock from a queue.
+  ///
+  /// \param Queue is a reference to a SYCL queue to which
+  ///        the mock redefinitions will apply
+  explicit PiMock(cl::sycl::queue &Queue)
+      : PiMock(std::make_shared<cl::sycl::platform>(
+            Queue.get_device().get_platform())) {}
+
+  /// Constructs PiMock from a reference to a SYCL platform instance.
+  ///
+  /// A new plugin will be stored into the platform instance, which
+  /// will no longer share the plugin with other platform instances
+  /// within the given context. A separate platform instance will be
+  /// held by the PiMock instance.
+  ///
+  /// \param OriginalPlatform is a reference to a SYCL platform
+  explicit PiMock(cl::sycl::platform &OriginalPlatform)
+      : PiMock(std::make_shared<cl::sycl::platform>(OriginalPlatform)) {}
+
+  /// Constructs PiMock from a pointer to a SYCL platform instance.
+  ///
+  /// A new plugin will be stored into the platform instance, which
+  /// will no longer share the plugin with other platform instances
+  /// within the given context.
+  ///
+  /// \param OriginalPlatformPtr is a pointer to a SYCL platform
+  explicit PiMock(std::shared_ptr<cl::sycl::platform> OriginalPlatformPtr) {
```

Comment:
Why is it `shared_ptr`?

Reply:
It's a leftover from an earlier, I-can't-remember-why-`unique_ptr`-wasn't-suitable approach. I've checked that `unique_ptr` works fine now, will update.

Reply:
I’m actually wondering why you construct pointer of platform at all. In a nutshell, platform is a shared_ptr-like object. It follows pimpl idiom. Why another level of indirection?

Reply:
There are a couple layers to the answer:
1. Whether `PiMock` should store any SYCL object handle at all (obviously, the one to `PiPlugin` is out of question). We might as well treat `PiMock` as a plain API over this `PiPlugin` handle and remove all complexity related to storing anything besides.
2. That said, I can imagine the scenario where the code user would want to pass a mocked object to another function/class/etc., _wherein to continue mocking_. They might end up having to pass the `PiMock` instance as well, unless it provides something like `getPlatform()`, which in turn requires `PiMock` to store some SYCL object handle.

If **2** is valid, the indirection could be minimized either by storing a `platform_impl` or a `plugin` handle instead, or providing a helper `pair`/`tuple` like object to enclose both the SYCL object and the `PiMock` instance.

What's your take here? Should we go down 1, 2.1, 2.2, "do nothing" or do something else?

Reply:
@AGindinson I'd probably go with simply storing `platform` object. Whenever you return it by value, the refcount of underlying `shared_ptr` is increased. So, you basically have the same platform, and you still can modify it freely. But you remove the complexities of storing one more `shared_ptr` inside mock platform.

Reply:
Fine. I'll adjust my current local version from `unique_ptr` to "by value" then.